### PR TITLE
Reusepython

### PR DESCRIPTION
**Pull request by peak3d**
## Description
Try to reuse the last used python thread / interpreter to avoid module reloading

- If an Python addon is invoked (yes, only addons, no script / no job), the thread (together with its Invoker instance) is saved and kept open.
- If the next python call requests the same addon script, the last saved thread is used instead opening a new thread + interpreter.
- If the user changes the addon, the last saved thread is freed and a new thread / interprter instanciated for the new addon script call.

## Motivation and Context
Launching each script call with a new Python interpreter instance takes much time. Reasons are mainly module reloading (own and foreign). Reusing the same Interpreter during navigating inside one addon reduces times significantly. On my test Addon from ~2 sec -> 100ms (per click) on ARM based device (Andoid and Linux)

## Types of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Notes
Using this commit changes behavior for addon developers. because script modules are not reread for every call, you have to switch to another addon and then back to your addon to see your changes made in python code

----------------------
If your submodules (all scripts which are not the default script) have global variables where comandline parameters are saved  (e.g. handle=sys.argv[1]), they will remain at their initial value (first call of addon). In order to work as expected, you either have to refresh them (passing from default.py to submodules) or use sys.arg[#] call in your submodules instead the global variable(s).

The first 2 parameter (url / pluginhandle) are now persistend over python sessions which are reused.
For the following parameters (action stuff) above mentioned note must be respected.

----------------------

This feature can be disabled in addon.xml file metadata section

```
<extension point="xbmc.addon.metadata">
  <reuselanguageinvoker>false</reuselanguageinvoker>
</extension>
```

### CODE REVIEW

**Comment by peak3d** [comment]
@zag2me you can add the attribute described in PR description to your addon.xml while developing

**Comment by candrews** [comment]
Instead of asking developers to set `<reuselanguageinvoker>false</reuselanguageinvoker>`, would it be possible to set up file monitoring so that when an addon .py is changed, the python thread using it is killed? This would allow (almost all of) the existing expectations to continue to be met.

File: xbmc/interfaces/python/PythonInvoker.cpp [hunk 1]
```diff
@@ -367,36 +367,41 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
...
+        break;
 
-    s = m_threadState->interp->tstate_head;
+      if (old != s)
+      {
+        CLog::Log(LOGINFO, "CPythonInvoker(%d, %s): waiting on thread %" PRIu64, GetId(), m_sourceFile.c_str(), (uint64_t)s->thread_id);
+        old = s;
+      }
+
+      lock.Leave();
```
Comment (jimfcarroll) [inline xbmc/interfaces/python/PythonInvoker.cpp]: While I'm not sure if anything here can throw, you should use an exit guard (CSingleExit)

**Comment by peak3d** [comment]
@candrews IMO users developing on addons should be able to edit addon.xml.
If we see in future that there is need for this, we'll implement

**Comment by enen92** [comment]
@Rechi @peak3d this needs the "wiki-needed" label. For reference, should be added to https://kodi.wiki/view/Addon.xml

**Comment by hudokkow** [comment]
@enen92 you're a team member. 😉 
You can add labels and milestones.

**Comment by Lunatixz** [comment]
@peak3d I'm seeing slowdowns and issues with this addition... Tested a number of plugins using windows nightly... anything I can do to help debug?

**Comment by peak3d** [comment]
@Lunatixz can you provide steps to reproduce pls? This PR was intended to make things faster :-)
Beside this there are many changes currently regarding message handling / busy dialog, so question is if you can really adress this PR as cause?

**Comment by L2501** [comment]
this broke script.module.routing

**Comment by peak3d** [comment]
@L2501 from irst sight it should not break this script.
_init_ should be called on every python call from kodi and reading sys there is fine.

Can you provide more information?

**Comment by L2501** [comment]
it is similar to the plugin.program.AML bug? need to add sys.argv to the plugin call. It also broke the viaplay addon https://github.com/emilsvennesson/kodi-viaplay/blob/master/default.py that uses a similar structure

**Comment by Lunatixz** [comment]
@peak3d, you are correct, there are a few commits I lost track of... I'll investigate and seek out the issue. THX

**Comment by zach-morris** [comment]
I can confirm this broke script.module.routing in [my addon](https://github.com/zach-morris/plugin.program.iagl).  If I add
```
<reuselanguageinvoker>false</reuselanguageinvoker>
```
to the addon.xml it works as expected.  I can't figure out specifically how it broke it though...

**Comment by peak3d** [comment]
@zach-morris pls. provide a debug log with a short / understandable description what "broke"

**Comment by Rechi** [comment]
@peak3d I posted a fix for the specific addon already at https://github.com/tamland/kodi-plugin-routing/issues/5 and submitted a PR (https://github.com/tamland/kodi-plugin-routing/pull/6) for updating the documentation.

**Comment by zach-morris** [comment]
Thanks.  That fixed it.

@peak3d, there is nothing in the logs to point to.  But given Rechi's documented example change, I get it now.
Here's what the documentation used to show:
```
import routing
import sys
from xbmcgui import ListItem
from xbmcplugin import addDirectoryItem, endOfDirectory

plugin = routing.Plugin()

@plugin.route('/')
def index():
    addDirectoryItem(plugin.handle, plugin.url_for(show_category, "one"), ListItem("Category One"), True)
    addDirectoryItem(plugin.handle, plugin.url_for(show_category, "two"), ListItem("Category Two"), True)
    endOfDirectory(plugin.handle)

@plugin.route('/category/<category_id>')
def show_category(category_id):
    addDirectoryItem(plugin.handle, "", ListItem("Hello category %s!" % category_id))
    endOfDirectory(plugin.handle)

if __name__ == '__main__':
    plugin.run()
```

Running this example would result in the addon showing the '/' route regardless of what was selected.

Changing the run call to
```
    plugin.run(sys.argv)
```
Resolves the issue

**Comment by tamland** [comment]
It was made clear months ago that this would break standard python. Who agreed to this? This should be reverted, or at the very least only be turned on when explicitly enabled by the add-on author.

**Comment by peak3d** [comment]
@tamland why? Its working now exactly in the way we always tell addon devs how our python engine works: Router is the thing which should be kept small, because other modules are not reloaded / refreshed.

If addon devs modify sys.args (for whatever reasin), it a.) simply changeable and b.) wrong from implementation POV.

Why not simply fix the issue and be happy that python is now much faster? Have you ever tried a bigger python addon on an rbpi?

**Comment by peak3d** [comment]
- module caching is a python functionality, I would really wonder if this would change in 3.x
- the PR was a.) field tested in LE, and b.) IMO long enough open before merged.
- Its not a workaround, its using a functionality provided by python.

For overwriting sys.args: this was the case in one of the posts above your comment.
There is a trac tickets about this.

I would never done this change if it would something what is not provided from python itself, but module caching is a core functiononality and we simply use it.

I'll support the few addon devs which have issues to get it right.

**Comment by tamland** [comment]
The python modules cache is *within the same process*. That's completely different. If I run a python program two times, sys.argv from the first does not leak into the other.

**Comment by peak3d** [comment]
> sys: This module provides access to some variables used or maintained by the interpreter and to functions that interact strongly with the interpreter. It is always available.

We have separate interpreter instances for each addon call, therefore from theory we have separate sys modules with separate sys parameters.

**Comment by Lunatixz** [comment]
@peak3d I agree with @tamland this commit breaks the routing for a number of plugins. Having developers enable path optimization is a better alternative then opting out.  In my personal case over 50 of my plugins in Kodis repo broke with this PR 👎 .

**Comment by peak3d** [comment]
@Lunatixz sorry, but really hard to deal with "broke my addon" coments.
Could you point to one of your addons which "broke"?

**Comment by Lunatixz** [comment]
@peak3d I'll pick one example, but you can take your pick https://github.com/Lunatixz/KODI_Addons/tree/master/plugin.video.looper
Inorder to fix my plugins i'll have to either push an updated to exclude your commit; or correct the path like below: 
https://github.com/Lunatixz/KODI_Addons/tree/master/plugin.video.wired

Just seems like it would have been easier to have developers enable support; rather then disable it...

**Comment by peak3d** [comment]
@Lunatixz : https://github.com/xbmc/xbmc/pull/14136

Maybe you want to remove the globals in your addons sooner or later.
This is not good python practise, you will get tons of "why's" if you google for "python globals".

https://github.com/Lunatixz/KODI_Addons/blob/master/plugin.video.wired/resources/lib/wired.py#L28-L47
is the onlyl issue, such initialization should be handled in functions, or if you are familiar with classes, in a python class (which contains __init__ constructor).
Many ways lead to rome, but the way using globals is the only one wich will fail (generally in python, not only in kodi python)

**Comment by Sorien** [comment]
hmm, are you familiar with fact that almost every addon tutorial i can find is using such globals even official sources like https://kodi.wiki/view/HOW-TO:Script_addon ;)

**Comment by peak3d** [comment]
@Sorien Thanx for pointing to the page,

If you have read carefully what I have written in PR and other comments, you may have noticed, that for router script (default.py) nothing has chaged. You can use globals there without any change. This is the reason why most of the scripts still run.

But: If you now decide do use submodules (scripts in resources lib), you most probably do this because you want to push performance. What we told users in the past (use submodules to speed up your addon), was not working correctly, but it is with the original PR.
But you have to simply follow some simple, easy to manage guidelines. This is basically: Use functions in your submodules, or use classes which are instanciated from router.
100% of the python modules are working fine, because they are build up as a lib. Only some koi addons do not work, because they are wrong by basic design.

Long story short: documentation in your linked article is fine, it does not touch self written submodules.

Suggestion to all of the ppl. who mislike this PR: Buy a rbpi, install LibreELEC, and play around with your addon. Testing addons on a Intel i7 does not cover kodi userbase.

**Comment by peak3d** [comment]
@Blackclaws: https://github.com/Sandmann79/xbmc/pull/142

**Comment by Lunatixz** [comment]
@peak3d Thanks for the info... I understand the methodology on why a programmer would want to avoid using globals. However in python/simple structure programs; globals are not that evil :)
Case in point ~75% of python modules in pypi use global variables =)

Anyway, off tangent... I hope you can see how your recent commit may benefit from opting to enable; rather then disable the path "optimization".

**Comment by romanvm** [comment]
> Buy a rbpi, install LibreELEC, and play around with your addon. 

I do have a Pi2 with LibreELEC as my primary Kodi system and still don't like all this idea of reusing an interpreter instance because it breaks expected behavior of the Python runtime. Usually, in case of plugins network latencies and server response times much bigger than delays because of module imports, especially if you organize your code properly to avoid bytecode recompiling at each strartup. So I'm with @Lunatixz and @tamland, and think that this should be disabled by default. If some addon does have the issue of slow startup because it imports tons of external modules, this feature can be enabled for this very addon, meaning it should be used only on case-by-case basis, but not by default.

**Comment by peak3d** [comment]
@romanvm Its not bytecode compiling, its IO.

**Comment by romanvm** [comment]
Still I have the feeling that this PR breaks YAGNI (you ain't gonna need it) principle, meaning that for most plugins performance improvement will be negligent compared to other factors like server response times. I admit that valid use cases for this probably exist, but there shouldn't be that many to justify the breaking change.

BTW, the current best practice for organizing addon code is to move everything to modules, leaving only a small "launcher" part in the main script, so that unchanged modules won't be re-compiled on each startup. Do I understan'd correctly that this PR makes this best practice obsolete, requiring to put everything in the main script again?

**Comment by peak3d** [comment]
@romanvm this PR is absolutely and only a benefit for using submodules and having the launcher as small as possible.
Without this PR its meaningless (you can put everything into one file, its most probably lots faster than using submodules in the implementation without the PR)

Only restriction is, that you treaten your submodules as submodules, and not as cosmetic include files (code separation)

**Comment by Gujal00** [comment]
Why is this PR marked as `Improvement (non-breaking change which improves existing functionality)`
And then in the Note say `Using this commit changes behavior for addon developers` The two are contradicting each other.
At least mark it correctly as `Breaking change (fix or feature that would cause existing functionality to change)` as it affects every other addon there is
